gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcApiExceptionFactory.java

```diff
@@ -67,8 +87,76 @@ ApiException create(Throwable throwable) {
     }
   }
 
-  private ApiException create(Throwable throwable, Status.Code statusCode) {
+  private ApiException create(Throwable throwable, Status.Code statusCode, Metadata metadata) {
     boolean canRetry = retryableCodes.contains(GrpcStatusCode.grpcCodeToStatusCode(statusCode));
-    return ApiExceptionFactory.createException(throwable, GrpcStatusCode.of(statusCode), canRetry);
+    if (metadata == null) {
+      return ApiExceptionFactory.createException(
+          throwable, GrpcStatusCode.of(statusCode), canRetry);
+    }
+    byte[] bytes = metadata.get(Metadata.Key.of(ERROR_DETAIL_KEY, Metadata.BINARY_BYTE_MARSHALLER));
+    if (bytes == null) {
+      return ApiExceptionFactory.createException(
+          throwable, GrpcStatusCode.of(statusCode), canRetry);
+    }
+
+    ErrorDetails.Builder errorDetailsBuilder = ErrorDetails.builder();
+    try {
+      com.google.rpc.Status status = com.google.rpc.Status.parseFrom(bytes);
+      List<String> detailsStringList = new ArrayList<>();
+      for (Any detail : status.getDetailsList()) {
+        Message unpack = unpack(detail);
+        if (unpack instanceof ErrorInfo) {
+          ErrorInfo errorInfo = (ErrorInfo) unpack;
+          errorDetailsBuilder.setDomain(errorInfo.getDomain());
+          errorDetailsBuilder.setReason(errorInfo.getReason());
+          errorDetailsBuilder.setErrorInfoMetadata(errorInfo.getMetadataMap());
+        }
+        detailsStringList.add("type: " + detail.getTypeUrl() + "\n" + unpack.toString());
+      }
+      errorDetailsBuilder.setDetailsStringList(detailsStringList);
+    } catch (InvalidProtocolBufferException | ClassNotFoundException e) {
+      // probably don't do anything?
+      e.printStackTrace();
```

Comment:
Please do not print anything. It is a very core library, it should not contaminate stdout in any way.

In terms of how to react to these exceptions: please consider either throwing exception, or putting in the resultant ErrorDetails some information indicating that there was an error while trying to parse error (^2 error =) and store the raw error details data in the ErrorDetails object instead of unpacked ones.

Reply:
Removed the print stack trace, this was for debugging purpose only. 
If unpacking error fails, I decided to not to do anything special and just return empty `ErrorDetails`, since the raw info would not be useful to users anyway. This behavior is also in sync with the implementations of other languages.